### ISSUE TITLE
Migrate to MessagePack as the de facto serialization format

### ISSUE THREAD

**Issue by marceline-cramer** [hearth-rs/hearth#253]
We've been using JSON, which bloats message size, is inefficient to encode and decode floats and integers, and requires conversion of blobs to base64 strings in order to have reasonable fast serialization. This kinda sucks!

A new message format that alleviates these problems should have the following properties:
- Binary format. Efficient data representation, especially for raw byte arrays.
- Schema-less. Can be made human-readable without the need for an external schema definition. Can be directly imported into a scripting environment as an object type.
- Simple. Easy to work with.

[MessagePack](https://msgpack.org/) satisfies all of these goals. No more base64! No more processes bottlenecked by ser/de!

I figure that along the way, since I don't foresee plain byte arrays as being used as messages too often, we could also migrate away from the `send`/`send_json` and `recv`/`recv_json` conventions and towards `send`/`send_raw` and `recv`/`recv_raw` instead, where the `_raw` functions work with plain byte arrays, and `send` and `recv` work with MessagePack.

To-do:
- add [rmp-serde](https://crates.io/crates/rmp-serde) as a workspace dep
- remove `serde_with` from schema and DON'T base64-encode anything
- add `rmp-serde` to runtime and use it to serialize events in `PubSub`, deserialize messages in `SinkProcess`, and serialize responses in `RequestResponseProcess`.
- swap out `serde_json` for `rmp-serde` in init
- swap out `serde_json` for `rmp-serde` in `run_wasm` example
- remove `serde_json` dep from workspace, schema, ctl, init, wasm, server, fs, and runtime.
- switch to `send`/`send_raw` and `recv`/`recv_raw` in guest
- swap out `serde_json` for `rmp-serde` in guest
- add `rmp-serde` to kindling workspace deps
- update all of the kindling crates that use `serde_json` to `rmp-serde`
- remove `serde_json` from kindling

Then, after #197:
- update `ByteVec` with custom `Deserialize` and `Serialize` implementations that read and write the serde bytes data type directly
- figure out what to do instead of `JsonAssetLoader`

**Comment by marceline-cramer** [comment]
Yeah, differentiating text and binary strings does sound like it'd come in handy, especially for scripting compatibility. I'll take a look at CBOR.

I've used Flatbuffers and looked at rkyv before but they don't fit our criteria for being schemaless. Both of these will definitely come in handy if we ever need very application-specific encoding formats where minimizing bandwidth is of utmost performance.

Thank you for your contribution! It's always good to see new faces. :)

**Comment by marceline-cramer** [comment]
CBOR looks fantastic!

[ciborium](https://crates.io/crates/ciborium) has serde support and a `Value` type that should integrate with future scripting languages really easily.

I was really concerned about how CBOR might encode very large integers, but [this](https://gist.github.com/sgsfak/04e5afb164fc0151afa180f314fd6329) makes it seem like there isn't a huge difference.

Yeah, let's use CBOR instead.

**Comment by marceline-cramer** [comment]
To answer your question both ways...

The reason why I decided on CBOR over MessagePack is:
- ciborium has a `Value` API, which is much easier to work with compared to `rmp`'s streaming encode/decode API
- CBOR is a web standard

The reason why I decided *so quickly* was because indecision can really paralyze these kinds of big sweeping design decisions, and even if there's some big mistake and we have to make all the same changes to the repository again with MessagePack instead or some other protocol, that's better than never attempting to migrate at all, or spending a lot of developer time studying and angsting over small differences between protocols. Since Hearth is such a small project right now, we have the benefit of really high agility in these kinds of decisions, so we should be taking advantage of that.

**Comment by marceline-cramer** [comment]
Could also ditch serde entirely and use this to reduce wasm binary sizes: https://docs.rs/minicbor/latest/minicbor/

**Comment by Meister1593** [comment]
When #278 closes I would like to try to dab into this, sounds fun and interesting

**Comment by marceline-cramer** [comment]
> When #278 closes I would like to try to dab into this, sounds fun and interesting

That would be fantastic, thank you so much! I figure that we should save potentially using `minicbor` for later down the line, if necessary.
